README.txt
= Hey!Watch - Video Encoding Web Service

Hey!Watch <http://heywatch.com> provides a simple and robust encoding plateform.
The service allows developers to access a fast, scalable and inexpensive web 
service to encode videos easier. The API can be easily integrated 
in any web or desktop applications.

The documentation of the API can be found at http://wiki.heywatch.com/API_Documentation


== Getting started

=== Transfer a video, encode it in ipod format and download the encoded video

 require 'heywatch'
 include HeyWatch

 Base::establish_connection! :login => 'login', :password => 'password'

 raw_video = Discover.create(:url => 'http://youtube.com/watch?v=SXcpNZCyQJw', :download => true) do |percent, total_size, received|
   puts "#{percent}%"
 end
 
 ipod_format = Format.find_by_name('iPod 4:3')
 encoded_video = Job.create(:video_id => raw_video.id, :format_id => ipod_format.id) do |percent|
   puts "#{percent}%"
 end

 puts "downloading {encoded_video.title}"
 path = encoded_video.download
 puts "video saved in {path}"

 
=== Upload a video from the disk, encode it with FTP auto transfer option

 raw_video = Video.create(:file => 'videos/myvideo.avi', :title => 'Funny video') do |percent, total_size, received|
   puts "#{percent}%"
 end

 Job.create :video_id => raw_video.id, :default_format => true, :ftp_directive => 'ftp://login:pass@host.com/heywatch_vids/'


=== Generate a thumbnail

  v = EncodedVideo.find(5400)
  v.thumbnail :start => 15, :width => 640, :height => 480    


=== Update your account

  account = Account.find
  account.update_attributes :ping_url_after_encode => 'http://yourhost.com/ping/encode'



== Integration in a rails application

This short HOWTO uses the ping options. So in your HeyWatch account, you must
configure all the ping URL (except transfer for this example).

Examples:

 ping_url_after_encode => http://myhost.com/ping/encode
 ping_url_if_error => http://myhost.com/ping/error


=== Config

In your config/environment.rb:

  require 'heywatch'
  HeyWatch::Base::establish_connection! :login => 'login', :password => 'passwd'


=== Item Model

 create_table "items", :force => true do |t|
   t.column "title",       :string,                          :null => false
   t.column "description", :text,                            :null => false
   t.column "created_at",  :datetime
   t.column "updated_at",  :datetime
   t.column "status",      :string,   :default => "working"
   t.column "url",         :string
   t.column "meta",        :text
   t.column "user_id",     :integer
 end

In app/models/item.rb:

 class Item < ActiveRecord::Base
   belongs_to :user
   serialize :meta
   after_create :convert

   # When HeyWatch send videos to your FTP,
   # they will be available to this HTTP url.
   def base_url
     "http://myhost.com/flv/#{self.id}/"
   end

   # Transfer the given URL and convert the video from 
   # this address in format 31 (flv format). When the encode
   # is done, the video will be sent to the specified FTP
   # with custom path.
   #
   # Note item_id which is a custom field.
   def convert        
     HeyWatch::Discover.create(
       :url              => self.url, 
       :download         => true, 
       :item_id          => self.id,
       :title            => self.title, 
       :automatic_encode => true,
       :format_id        => 31,
       :ftp_directive    => "ftp://login:passwd@myhost.com/flv/#{self.id}/"
     )
   end
 end


=== Ping Controller

In app/controllers/ping_controller.rb:

 class PingController < ApplicationController
   before_filter :find_item

   def encode
     @encoded_video = HeyWatch::EncodedVideo.find(params[:encoded_video_id])
     @full_url = @item.base_url + @encoded_video.filename
     @item.status = 'finished'
     @item.meta = {
       :url       => @full_url,
       :thumb     => @full_url + ".jpg",
       :size      => @encoded_video.specs["size"],
       :mime_type => @encoded_video.specs["mime_type"],
       :length    => @encoded_video.specs.video["length"]
     }
     @item.save
     @encoded_video.job.video.destroy # delete the raw video
     @encoded_video.destroy # delete the encoded video
     render :nothing => true
   end

   def error
     if params[:discover_id]
       error_msg = "No video link found"
     else
       error_msg = HeyWatch::ErrorCode[params[:error_code].to_i]      
     end
     @item.update_attributes :status => 'error'
     ItemMailer.deliver_error(error_msg, @item.user)
     render :nothing => true
   end

   private

   # item_id is a custom_field sent in Item#convert.
   # Thanks to it, we can track the item.
   def find_item
     @item = Item.find params[:item_id]
   end
 end


CHANGELOG
*0.0.1 (August 8, 2007)*

* First Official Release [Bruno Celeste]


test/test_helper.rb
require 'test/unit'
require File.dirname(__FILE__) + '/../lib/heywatch' unless defined?(HeyWatch)

VIDEO_URL = 'http://www.youtube.com/watch?v=eBYLyt9Ptrs'

def discover_and_download_video(url)
  Discover.create(:url => url, :download => true, :automatic_encode => false) {|percent, total, received| puts "#{percent}%" }
end

def get_credentials(type)
  filename = "#{ENV['HOME']}/.heywatch-#{type}"
  return File.read(filename).chomp if File.exists?(filename)
  nil
end

test/test_heywatch.rb
require File.dirname(__FILE__) + '/test_helper.rb'
include HeyWatch


class TestHeywatch < Test::Unit::TestCase
  def setup
    Base::establish_connection! :login => get_credentials(:login), :password => get_credentials(:password)
  end
  
  # Authentification
  
  def test_should_be_authorized
    assert_kind_of Account, Account.find
  end
  
  def test_should_raise_if_not_authorized
    assert_raise(NotAuthorized) { Auth::create 'fakelogin', 'fakepasswd' }
  end
  
  def test_should_raise_if_not_logged_in
    Base::disconnect!
    assert_raise(NotAuthorized) { Format.find(:first) }
  end
  
  # Finding
  
  def test_should_return_the_first_format
    assert_kind_of Format, Format.find(:first)
  end
  
  def test_should_return_format_with_id
    id = Format.find(:first).id
    assert_kind_of Format, Format.find(id)
  end
  
  def test_should_return_all_formats
    formats = Format.find(:all)
    assert_kind_of Array, formats
    assert formats.size > 0
  end
  
  def test_should_return_just_three_formats
    formats = Format.find(:all, :limit => 3)
    assert formats.size == 3
  end
  
  def test_should_return_all_ipod_formats
    formats = Format.find(:all, :conditions => {:name => /ipod/i})
    formats.each do |f|
      assert f.name =~ /ipod/i
    end
  end
  
  def test_should_return_all_formats_using_xvid_codec
    formats = Format.find(:all, :conditions => {:video_codec => 'xvid'})
    formats.each do |f|
      assert f.video_codec == 'xvid'
    end
  end
  
  def test_should_return_all_formats_having_width_less_than_176
    formats = Format.find(:all, :conditions => {:width => '< 176'})
    formats.each do |f|
      assert f.width < 176
    end
  end
  
  def test_should_find_by_name
    assert_equal 'Mobile 3GP', Format.find_by_name('Mobile 3GP').name
  end
  
  def test_should_find_all_by_video_codec
    formats = Format.find_all_by_video_codec('xvid')
    assert_kind_of Array, formats
    formats.each do |f|
      assert f.video_codec =='xvid'
    end
  end
  
  def test_should_raise_if_not_found
    assert_raise(ResourceNotFound) { Format.find(99999999) }
  end
  
  # Discover
  
  def test_should_create_discover
    discover = Discover.create :url => VIDEO_URL
    assert_kind_of Discover, discover
  end
  
  def test_should_create_discover_and_go_until_raw_video
    raw_video = discover_and_download_video VIDEO_URL
    assert_kind_of Video, raw_video
  end
  
  def test_should_raise_because_invalid_url
    assert_raise(RequestError) { Discover.create :url => 'invalid_url' }
  end
  
  def test_should_raise_because_no_video_found
    assert_raise(VideoNotFound) { discover_and_download_video 'http://google.com' }
  end
  
  # Job
  
  # Each of these tests consume 1 credit.
  
  #def test_should_create_job
  #  raw_video = discover_and_download_video VIDEO_URL
  #  job = Job.create :video_id => raw_video.id, :format_id => Format.find_by_name('Mobile 3GP').id
  #  assert_kind_of Job, job
  #end
  
  #def test_should_create_job_and_go_until_encoded_video
  #  raw_video = discover_and_download_video VIDEO_URL
  #  encoded_video = Job.create(:video_id => raw_video.id, :format_id => Format.find_by_name('Mobile 3GP').id) do |percent|
  #    puts "#{percent}%"
  #  end
  #  assert_kind_of EncodedVideo, encoded_video
  #end
  
  def test_should_raise_if_format_doesnt_exist
    raw_video = discover_and_download_video VIDEO_URL
    assert_raise(RequestError) { Job.create :video_id => raw_video.id, :format_id => 80000 }
  end
  
  def test_should_raise_if_video_doesnt_exist
    assert_raise(RequestError) { Job.create :video_id => 150, :format_id => 2 }
  end
  
  # Update and Delete
  
  def test_should_update_the_title
    raw_video = discover_and_download_video VIDEO_URL
    raw_video.update_attributes :title => 'mytitle'
    assert_equal 'mytitle', Video.find(raw_video.id).title
  end
  
  def test_should_delete_the_video
    raw_video = discover_and_download_video VIDEO_URL
    assert raw_video.destroy
  end
  
  def test_should_raise_if_video_doesnt_exist
    assert_raise(ResourceNotFound) { Video.destroy(99999999999) }
    assert_raise(ResourceNotFound) { Video.update(99999999999, :title => 'mytitle') }
  end
  
  # Browser
  
  def test_should_retry_connection_attempt_on_network_errors
    [Errno::EPIPE, Timeout::Error, Errno::EPIPE, Errno::EINVAL, EOFError].each do |network_exception|
      retrial = 0
      assert_raise(network_exception) {
        Browser.with_http {|http| retrial += 1; raise network_exception}
      }
      assert_equal 3, retrial
    end
  end
  
  # Hash and Array extension
  def test_attribute_hash_should_compare_to_hash
    assert_equal({:foo => {'bar' => 1}}, 
                  AttributeHash.new({:foo => AttributeHash.new({'bar' => 1})}))
                 
  end
  
  def test_attribute_hash_should_serialize_as_hash
    assert_equal({:foo => {'bar' => 1}}.to_yaml, 
                  AttributeHash.new({:foo => AttributeHash.new({'bar' => 1})}).to_yaml)
  end
  
  def test_response_collection_should_compare_to_array
    assert_equal [:nice,:nice],
                 ResponseCollection.new(2, :nice)
  end
  
  def test_response_collection_should_serialize_as_hash
    assert_equal [:nice,:nice].to_yaml,
                 ResponseCollection.new(2, :nice).to_yaml
  end
end

lib/heywatch.rb
#--
# Copyright (c) 2007 Bruno Celeste
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#++

require "rubygems"
require "xmlsimple"
require "cgi"

module HeyWatch
	Host = "heywatch.com" unless const_defined? :Host
	OutFormat = "xml" unless const_defined? :OutFormat
  Resources = %w{job format download discover encoded_video video account log} unless const_defined? :Resources

  # These are error codes you receive if you use ping options
  ErrorCode = {
                100 => "Unknown error",
                101 => "Unsupported audio codec",
                102 => "Unsupported video codec",
                103 => "This video cannot be encoded in this format",
                104 => "Wrong settings for audio",
                105 => "Wrong settings for video",
                106 => "Cannot retrieve info from this video",
                107 => "Not a video file",
                108 => "Video too long",
                109 => "The container of this video is not supported yet",
                110 => "The audio can't be resampled",
                201 => "404 Not Found",
                202 => "Bad address",
                300 => "No more credit available"
              } unless const_defined? :ErrorCode
              
  class NotAuthorized < RuntimeError; end
  class RequestError < RuntimeError; end
  class ResourceNotFound < RuntimeError; end
  class ServerError < RuntimeError; end
  class SessionNotFound < RuntimeError; end
  
  # Convert the XML response into Hash
  def self.response(xml)
    AttributeHash.new(XmlSimple.xml_in(xml, {'ForceArray' => false, 'keeproot' => false}))
  end
  
  # sanitize url
  def self.sanitize_url(url)
    return url.gsub(/[^a-zA-Z0-9:\/\.\-\+_\?\=&]/) {|s| CGI::escape(s)}.gsub("+", "%20")
  end
end

$: << File.dirname(File.expand_path(__FILE__))

require "heywatch/ext"
require "heywatch/version"
require "heywatch/browser"
require "heywatch/auth"
require "heywatch/base"
require "heywatch/encoded_video"
require "heywatch/video"
require "heywatch/account"
require "heywatch/download"
require "heywatch/discover"
require "heywatch/job"

lib/heywatch/ext.rb
unless String.method_defined?(:camelize)
  class String
    # convert "undercore_string" in "UnderscoreString"
    #
    # Method from Rails
    def camelize
      self.gsub(/\/(.?)/) { "::" + $1.upcase }.gsub(/(^|_)(.)/) { $2.upcase }
    end
  end
end

module HeyWatch
  class AttributeHash < Hash
    def initialize(constructor = {})
      if constructor.is_a?(Hash)
        super()
        update(constructor)
      else
        super(constructor)
      end
    end
    
    # Returns an exact copy
    def dup
      AttributeHash.new(self)
    end
    
    # Avoid keeping concrete type in stuff like serialization
    def to_yaml(*args)
      to_hash.to_yaml(*args)
    end
        
    # Convert to a Hash
    def to_hash
      Hash.new(default).merge(self)
    end
    
    # Replace "-" by "_" in all keys of the hash
    def underscore_keys!
      self.keys.each do |k|
        if self[k].is_a?(Hash)
          self[k] = AttributeHash.new(self[k]) 
          self[k] = self[k].underscore_keys!
        end
        self[k.to_s.gsub("-", "_")] = self[k]
        delete(k) if k.to_s =~ /-/
      end
      self
    end
    
    # Cast each value of the hash in the right format (integer, float, time, boolean)
    def type_cast!
      self.keys.each do |k|
        self[k] = self[k].type_cast! if self[k].is_a?(Hash)
        self[k] = self[k].to_i if self[k] =~ /^[0-9]+$/
        self[k] = self[k].to_f if self[k] =~ /^[0-9]+\.[0-9]+$/
        self[k] = Time.parse(self[k]) if self[k] =~ /[a-z]+ [a-z]+ [0-9]+ [0-9]{2}:[0-9]{2}:[0-9]{2} [a-z0-9\+]+ [0-9]{4}/i
        self[k] = true if self[k] == "true"
        self[k] = false if self[k] == "false"
        self[k] = nil if self[k] == {}
      end
      self
    end
    
    # The hash keys are available by calling the method name "key"
    #
    #  h = {"format_id" => 54, "video_id" => 12}
    #  h.format_id
    #  => 54
    def method_missing(m, *args)
      if has_key? m.to_s
        self[m.to_s]
      else
        super
      end
    end    
  end
  
  class ResponseCollection < Array
    def find_with_options(options={})
      collection = self
      if options[:conditions]
        options[:conditions].each_pair do |k,v|
          collection = collection.find_with_conditions(k => v)
        end
      end
      collection.limit(options[:limit]).order(options[:order]).include_heywatch_object(options[:include])
    end
    
    
    # Select HeyWatch#Base object containing the conditions
    #
    #   Format.find(:all).find_with_conditions(:owner => true)
    def find_with_conditions(conditions={})
      return self if conditions.nil? or conditions.empty?
      res = ResponseCollection.new
      self.each do |object|
        object.attributes.each_pair do |k,v|
          if conditions[k.to_sym]
            condition = []
            if not conditions[k.to_sym].is_a?(Regexp) and conditions[k.to_sym].to_s =~ /([<|>|<=|>=|==]+)[\s]*([0-9]+)$/
              conditions[k.to_sym].to_s.split(" and ").each do |c|
                c.match(/([<|>|<=|>=|==]+)[\s]*([0-9]+)$/)
                condition << "#{v} #{$1} #{$2}"
              end
              res << object if eval(condition.join(" and "))
            elsif conditions[k.to_sym].is_a?(Regexp)
              res << object if v =~ conditions[k.to_sym]
            else
              res << object if v == conditions[k.to_sym]
            end
          end
        end
      end
      return res
    end

    def limit(lim=nil)
      return self if lim.nil?
      self[0..lim-1]
    end

    def order(order_str=nil)
      return self if order_str.nil?
      field, sort = order_str.match(/([a-z0-9_]+)([\s+]*[ASC|DESC]*)/i).to_a[1..-1]
      sort = "ASC" if sort.nil? or sort.empty?
      begin
        self.sort do |x,y|
          case sort
          when /ASC/i then x.send(field) <=> y.send(field)
          when /DESC/i then y.send(field) <=> x.send(field)
          end
        end
      rescue
        self
      end
    end
    
    def include_heywatch_object(inc=nil)
      return self if inc.nil?
      self.each {|object| object.include_heywatch_object(inc)}
    end
    
    def to_array
      [] + self
    end
    
    def to_yaml(*args)
      to_array.to_yaml(*args)
    end
  end
end

lib/heywatch/browser.rb
require "net/http"

module HeyWatch
  # This class is used to request the Hey!Watch service.
  #
  #   Browser::get '/video', session
  #   Browser::post '/download', {:url => 'http://host.com/video.avi'}, session
  #   Browser::put '/encoded_video/54000', {:title => 'new title'}, session
  #   Browser::delete '/encoded_video/54000', session
  class Browser
    class << self
      def with_http(&block)
        attempts ||= 1
        cnx = Net::HTTP.new(Host)
        cnx.open_timeout = 10
        cnx.read_timeout = 10
        cnx.start &block
      rescue Errno::EPIPE, Timeout::Error, Errno::EPIPE, Errno::EINVAL, EOFError
        attempts += 1
        attempts <= 3 ? retry : raise
      end
      
      # Raise when code response != 2xx
      def raise_if_response_error(res)
        code = res.response.code.to_i
        message = res.response.message
        return if code.to_s =~ /^2/
        
        raise RequestError, HeyWatch::response(res.body).content if code == 400
        raise NotAuthorized, message if code == 401
        raise ResourceNotFound, message if code == 404
        raise ServerError, message if code == 500
      end

      def header(session=nil) #:nodoc:
        h = {}
        h.merge!({"Cookie" => "_session_id=#{session};"}) if session
        h.merge!({"User-Agent" => "Hey!Watch ruby API - #{VERSION::STRING}"})
      end
      
      # Login to Hey!Watch service. Return the session ID.
      #
      # You should not use it directly, use Auth#create instead
      #
      #  Browser::login 'login', 'password'
      def login(login, password) #:nodoc:
        res = Browser::post("/auth/login", :login => login, :password => password)
        return res["Set-cookie"].match(/_session_id=(.*);/i)[1].to_s
      end
      
      # GET on path
      def get(path, session=nil)
        path += ".#{OutFormat}" unless path.include? "."
        res = with_http {|http| http.get(path, header(session))}
        raise_if_response_error(res)
        res
      end

      # POST on path and pass the query(Hash)
      def post(path, query={}, session=nil)
        res = with_http {|http| http.post(path, query.merge(:format => OutFormat).to_a.map{|x| x.join("=")}.join("&"), self.header(session))}
        raise_if_response_error(res)
        res
      end
      
      # PUT on path and pass the query(Hash)
      def put(path, query={}, session=nil)
        req = Net::HTTP::Put.new(path, header(session))
        req.form_data = query.merge(:format => OutFormat)
        res = with_http {|http| http.request(req) }
        raise_if_response_error(res)
        true
      end
      
      # DELETE on path
      def delete(path, session=nil)
        res = with_http {|http| http.delete(path+"."+OutFormat, header(session))}
        raise_if_response_error(res)
        true
      end
      
      def post_multipart(path, attributes={}, session=nil) #:nodoc:
        file = attributes.delete(:file)
        params = [file_to_multipart("data", File.basename(file),"application/octet-stream", File.read(file))]
        attributes.merge("format" => OutFormat).each_pair{|k,v| params << text_to_multipart(k.to_s, v.to_s)}

        boundary = '349832898984244898448024464570528145'
        query = params.collect {|p| '--' + boundary + "\r\n" + p}.join('') + "--" + boundary + "--\r\n"
        res = with_http {|http| http.post(path, query, header(session).merge("Content-Type" => "multipart/form-data; boundary=" + boundary))}
        raise_if_response_error(res)
        res
      end

      def text_to_multipart(key,value) #:nodoc:
        return "Content-Disposition: form-data; name=\"#{CGI::escape(key.to_s)}\"\r\n" + 
               "\r\n" + 
               "#{value}\r\n"
      end

      def file_to_multipart(key,filename,mime_type,content) #:nodoc:
        return "Content-Disposition: form-data; name=\"#{CGI::escape(key.to_s)}\"; filename=\"#{filename}\"\r\n" +
               "Content-Transfer-Encoding: binary\r\n" +
               "Content-Type: #{mime_type}\r\n" + 
               "\r\n" + 
               "#{content}\r\n"
      end
    end
  end
end


lib/heywatch/base.rb
module HeyWatch 
  # You can connect to the Hey!Watch service permanently
  #
  #  Base::establish_connection! :login => "login", :password => "password"
  #
  # You can access all the resources and manipulate the objects like in ActiveRecord
  #
  #  EncodedVideo.find(:all, :conditions => {:title => /bad video/i}).each do |v|
  #    v.destroy
  #  end
  #
  #  Format.create :name => "my new format", :video_codec => "mpeg4", ...
  #  
  #  f = Format.new
  #  f.container = "avi"
  #  f.audio_codec = "mp3"
  #  ...
  #  f.save
  #
  #  v = Video.find(15).update_attributes :title => "My edited title"
  #  puts v.title
  #
  #  Format.find_by_name("iPod 4:3")
  #  
  #  Format.find_all_by_name /ipod/i
  class Base
    attr_reader :attributes, :errors
    
    class << self
      def session=(session) #:nodoc:
        @session = session
      end

      def session #:nodoc:
        (@session || Base.session) rescue nil
      end
      
      # Establish the connection for all your session
      def establish_connection!(options={})
        Base.session = Browser::login(options[:login], options[:password])
      end

      def disconnect!
        Base.session = nil
      end
      
      def method_missing(m, *args) #:nodoc:
        if m.to_s =~ /find_by_(.*)/
          find(:first, :conditions => {$1.to_sym => args.first}) rescue nil
        elsif m.to_s =~ /find_all_by_(.*)/
          find :all, {:conditions => {$1.to_sym => args.first}}.merge(args[1]||{})
        else
          raise NoMethodError, "undefined method `#{m.to_s}' for #{self}"
        end
      end

      def path #:nodoc:
        return @path if @path
        "/"+self.to_s.split("::").last.
        gsub(/([A-Z]+)([A-Z][a-z])/,'\1_\2').
        gsub(/([a-z\d])([A-Z])/,'\1_\2').
        tr("-", "_").
        downcase
      end
      
      # if the path of the resource is not standard, you can set it manually.
      #
      #   class Log < Base
      #      self.path = "/user_logs"
      #   end
      def path=(path) #:nodoc:
        @path = path
      end

      # Find objects
      # 
      # Arguments:
      #
      # * <tt>ID</tt> ID of the object
      # * <tt>:first</tt> retrieve the first result
      # * <tt>:all</tt> retrieve all the results
      # 
      # Options:
      #   
      # * <tt>:conditions</tt> {:fieldname => "keyword"} or {:fieldname => /[0-9]+/}
      # * <tt>:order</tt> to sort the result
      # * <tt>:limit</tt> limit the number of result to return
      # * <tt>:include</tt> fetch the object to include
      #
      #     Format.find(:all, :conditions => {:name => /mobile/i})
      #     Format.find(:all, :conditions => {:width => '<320'})
      #     Job.find(:all, :include => "encoded_video", :order => "created_at DESC", :limit => 3)
      #     Video.find :first
      #     Download.find(5)
      def find(*args)
        scope, options = args     
        options ||= {}
        case scope
          when :all   then find_every(options)
          when :first then find_every(options).first
          else             find_single(scope, options)
        end
      end
      
      private
      
      def find_every(options) #:nodoc:
        collection = ResponseCollection.new
        res = HeyWatch::response(Browser::get(path, session).body)
        return collection if res.empty?
        
        [res[res.keys.first]].flatten.each {|object| collection << new(object)}
        collection.find_with_options(options)
      end
      
      def find_single(arg, options) #:nodoc:
        new(HeyWatch::response(Browser::get(path+"/"+arg.to_s, session).body)).include_heywatch_object(options[:include])
      end
      
      public
        
      # Create the object
      #
      #   Download.create :url => "http://host.com/video.avi"
      def create(attributes={})
        new(HeyWatch::response(Browser::post(path, attributes, session).body))
      end
      
      # Update the object passing its ID
      #
      #   EncodedVideo.update 15, :title => "my title"
      def update(id, attributes={})
        Browser::put(path+"/"+id.to_s, attributes, session)
      end
      
      # Destroy the object passing its ID
      #
      #   Format.destroy 12
      def destroy(id)
        Browser::delete(path+"/"+id.to_s, session)
      end
      
      # Destroy all the objects
      #
      #   Video.destroy_all
      def destroy_all
        find(:all).each do |object|
          object.destroy
        end
      end

      # Count request
      #
      # Accept :conditions like in Base#find
      def count(field=nil, options={})
        find(:all, options).size
      end
    end
    
    # Instanciate a new object
    #
    #   Format.new :name => "test format", :sample_rate => 24000
    def initialize(attributes={})
      @attributes = AttributeHash.new(attributes)
      @attributes.underscore_keys!
      @attributes.type_cast!
    end

    def include_heywatch_object(objects=nil) #:nodoc:
      return self if objects.nil?
      objects = objects.to_s.split(",") if objects.is_a?(String) or objects.is_a?(Symbol)
      objects.each do |ob|
        begin
          self.instance_variable_set "@#{ob}", (self.send(ob) rescue nil)
          self.instance_eval "attr_reader #{ob.to_sym}"
        rescue
        end
      end
      self
    end

    # Save the object. 
    #
    # If the object doesn't exist, it will be created, otherwise updated.
    # If an error occurred, @errors will be filled. This method doesn't raise.
    def save
      begin
        save!
        true
      rescue => e
        @errors = e.to_s
        false
      end
    end
    
    # Save the current object
    #
    # Raise if an error occurred. Return self.
    def save!
      if new_record?
        self.class.create(@attributes)
      else
        update_attributes(@attributes)
      end
      self
    end

    def id #:nodoc:
      @attributes["id"].to_i
    end
    
    # Update the object with the given attributes
    #
    #   Video.find(10).update_attributes :title => "test title"
    def update_attributes(attributes={})
      if self.class.update(id, attributes)
        reload
        true
      end
    end
    
    # Destroy the object
    #
    #   Video.find(56).destroy
    def destroy
      self.class.destroy(id)
    end
    
    # Reload the current object
    #
    #   j = Job.find(5400)
    #   j.reload
    def reload
      unless new_record?
        @attributes = self.class.find(self.id).attributes
      end
      self
    end
    
    def new_record? #:nodoc:
      @attributes["id"].nil?
    end
    
    def method_missing(m, *args) #:nodoc:
      method_name = m.to_s
      case method_name[-1..-1]
      when '='
        @attributes[method_name[0..-2]] = *args.first
      when '?'
        @attributes[method_name[0..-2]]
      else
        if instance_variables.include?("@#{m.to_s}")
          eval("@#{m.to_s}")
        else
          if object_id = @attributes[m.to_s+"_id"] # belongs_to
            klass = HeyWatch::const_get(m.to_s.camelize)
            klass.session = self.class.session
            klass.find(object_id)
          else
            @attributes[m.to_s] rescue nil
          end
        end
      end
    end
  end

  Resources.each do |k| 
    HeyWatch.module_eval(%{class #{k.to_s.camelize} < Base; end})
  end
end
